From 397311868bde9b25beebf9b4df62e0e86546b49a Mon Sep 17 00:00:00 2001
From: Philippe Normand <philn@igalia.com>
Date: Mon, 7 Dec 2015 16:33:49 +0100
Subject: [PATCH] qtdemux: fallback to parsed duration reporting if it's
 available

In some cases the upstream element handles the duration query but
doesn't provide a valid value. In that situation and if qtdemux
already has information about the media duration then set it as query
result.

https://bugzilla.gnome.org/show_bug.cgi?id=759126
---
 gst/isomp4/qtdemux.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/gst/isomp4/qtdemux.c b/gst/isomp4/qtdemux.c
index bcec740..fa5b881 100644
--- a/gst/isomp4/qtdemux.c
+++ b/gst/isomp4/qtdemux.c
@@ -890,10 +890,13 @@ gst_qtdemux_handle_src_query (GstPad * pad, GstObject * parent,
 
       gst_query_parse_duration (query, &fmt, NULL);
       if (fmt == GST_FORMAT_TIME) {
+        guint64 duration;
         /* First try to query upstream */
         res = gst_pad_query_default (pad, parent, query);
-        if (!res) {
-          guint64 duration;
+        gst_query_parse_duration (query, NULL, &duration);
+
+        /* Fallback to parsed duration if upstream query failed or didn't return a valid duration */
+        if (!res || (duration == GST_CLOCK_TIME_NONE)) {
           if (gst_qtdemux_get_duration (qtdemux, &duration) && duration > 0) {
             gst_query_set_duration (query, GST_FORMAT_TIME, duration);
             res = TRUE;
-- 
2.6.2

